soroban-env-host/src/crypto/bls12_381.rs

```diff
@@ -0,0 +1,665 @@
+        msg: &str,
+    ) -> Result<(), HostError> {
+        self.as_budget()
+            .bulk_charge(ContractCostType::Bls12381EncodeFp, units_of_fp, None)?;
+        element.serialize_uncompressed(buf).map_err(|_e| {
+            self.err(
+                ScErrorType::Crypto,
+                ScErrorCode::InternalError,
+                format!("bls12-381: unable to serialize {msg}").as_str(),
+                &[],
+            )
+        })?;
+        Ok(())
+    }
+
+    fn validate_point_encoding<const EXPECTED_SIZE: usize>(
+        &self,
+        bytes: &[u8],
+        msg: &str,
+    ) -> Result<(), HostError> {
+        // validate input bytes length
+        if EXPECTED_SIZE == 0 || bytes.len() != EXPECTED_SIZE {
+            return Err(self.err(
+                ScErrorType::Crypto,
+                ScErrorCode::InvalidInput,
+                format!("bls12-381 {msg}: invalid input length to deserialize").as_str(),
+                &[
+                    Val::from_u32(bytes.len() as u32).into(),
+                    Val::from_u32(EXPECTED_SIZE as u32).into(),
+                ],
+            ));
+        }
+        // validated encoded flags. The most significant three bits encode the flags,
+        // i.e. `byte[0] == [compression_flag, infinity_flag, sort_flag, bit_3, .. bit_7]`
+        // - the compression_flag should be unset
+        // - the infinity_flag should be set **only if** rest of bits are all zero
+        // - the sort_flag should be unset
+        let flags = 0b1110_0000 & bytes[0];
+        match flags {
+            0b0100_0000 => {
+                // infinite bit is set, check all other bits are zero
+                let mut expected_bytes = [0; EXPECTED_SIZE];
+                expected_bytes[0] = flags;
+                if bytes != expected_bytes {
+                    Err(self.err(ScErrorType::Crypto, ScErrorCode::InvalidInput, format!("bls12-381 {msg} deserialize: infinity flag (bit 1) is set while remaining bits are not all zero").as_str(), &[]))
+                } else {
+                    Ok(())
+                }
+            },
+            0b0000_0000 => Ok(()), // infinite bit is unset
+            _ => Err(self.err(ScErrorType::Crypto, ScErrorCode::InvalidInput, format!("bls12-381 {msg} deserialize: either compression flag (bit 0) or the sort flag (bit 2) is set, while the input should be encoded uncompressed").as_str(), &[]))
+        }
+    }
+
+    pub(crate) fn metered_check_point<P: SWCurveConfig>(
+        &self,
+        pt: Affine<P>,
+        ty: ContractCostType,
+    ) -> Result<Affine<P>, HostError> {
+        self.charge_budget(ty, None)?;
+        // performs following checks 1. point belongs to the curve 2. point
+        // belongs to the correct subgroup
+        pt.check().map_err(|_| {
+            self.err(
+                ScErrorType::Crypto,
+                ScErrorCode::InvalidInput,
+                "bls12-381 G1 affine deserialize: invalid point",
```

Comment:
The error message hardcodes G1, while this method is used by both G1 and G2.

Reply:
fixed